Fix for GEODE-3292

jujoramos opened the pull request:
Embedded PULSE Connection Failure When jmx-manager-bind-address != localhost

- ManagementAgent sets the 'pulse.host' system property whenever a 'jmx-manager-bind-address' is configured.
- PULSE gets the hostname of the local JMX Manager through the 'pulse.host' system property, reverting to 'localhost' by default.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

jaredjstewart commented:
Hi Juan,

Thanks for your contribution!  Your changes look good to me, but our [Criteria for Code Submissions](https://cwiki.apache.org/confluence/display/GEODE/Criteria+for+Code+Submissions) require having tests for the behavior that is being changed.  In this case, it would be good to add a test to make sure that GEODE-3292 has been fixed and will not reoccur.  Typically I would point you towards an existing test to use as a guide, but since your changes appear to be in an area of the code that is not particularly well-tested, let me know if you don't feel confident in writing such a test and I can help.

jujoramos commented:
Hey @jaredjstewart,

Thanks for looking at this.
I'll write the tests, no worries at all.
Cheers.

jujoramos commented:
Hello @jaredjstewart,

Just pushed the requested changes.

- Renamed `PulseVerificationTest` to `PulseConnectivityTest`.
- Removed `loginWithIncorrectPassword` test method, it is the same as the one in `PulseSecurityTest`.
- Added integration tests to check the connectivity between PULSE and the local jmx-manager, both when the default and non-default bind-address is used.
- Added `PulseAppListenerTest` unit test to verify that the repository is correctly configured by the `PulseAppListener` when PULSE is running in embedded mode.

Cheers.

jaredjstewart commented:
Thanks @jujoramos, this looks good to me.  Do you have any thoughts @jinmeiliao?

jinmeiliao commented on geode-assembly/src/test/java/org/apache/geode/tools/pulse/PulseConnectivityTest.java:
```diff
@@ -0,0 +1,110 @@
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+
+@RunWith(Enclosed.class)
```
I am not sure I've seen this pattern before. It looks like you are adding another set of test to the original PulseVerificationTest. I believe another separate test class would read better. I would leave the original test alone (since it has a test that uses httpClient to try to connect with a wrong password).

jinmeiliao replied:
the test you removed is slightly different from the one in PulseSecurityTest, there it's using a securityManager, but here, it doesn't, it's making sure the default username/password (admin/admin) works.

jujoramos replied:
Yup, you're right, I didn't see the security manager within the rule and just looked at the method implementation. The login test doesn't seem to be functionality related to the connectivity test, that's why I removed it completely and aggregated the tests in a single class with two inner classes.

jinmeiliao commented on geode-assembly/src/test/java/org/apache/geode/tools/pulse/PulseConnectivityTest.java:
```diff
@@ -0,0 +1,110 @@
+    public EmbeddedPulseRule pulse = new EmbeddedPulseRule();
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
```
if you use new LocatorStarterRule().withAutoStart(), you don't need this @BeforeClass block anymore.

jujoramos replied:
True, but I'm not using `withAutoStart()` for a good reason: the two inner classes need to setup the locator differently (the `jmx-manager-bind-address` value changes). 
I thought it's easier to read a single functional test class annotated with `@RunWith(Enclosed.class)` internally split in different inner classes than write a new test class to test exactly the same but changing a single locator property.

jinmeiliao commented on geode-assembly/src/test/java/org/apache/geode/tools/pulse/PulseConnectivityTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
```
this is not needed. The rule guarantees that this is called.

jinmeiliao commented on geode-assembly/src/test/java/org/apache/geode/tools/pulse/PulseConnectivityTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
```
Not needed.

jinmeiliao commented on geode-assembly/src/test/java/org/apache/geode/test/dunit/rules/EmbeddedPulseRule.java:
```diff
@@ -38,6 +38,12 @@ protected void before() throws Throwable {
     repository.setHost("localhost");
   }
 
+  public void useJmxManager(String jmxHost, int jmxPort) {
```
+1, very nice addition

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/management/internal/ManagementAgent.java:
```diff
@@ -270,6 +271,7 @@ private void startHttpService(boolean isServer) {
           }
 
           System.setProperty(PULSE_EMBEDDED_PROP, "true");
+          System.setProperty(PULSE_HOST_PROP, "" + config.getJmxManagerBindAddress());
```
+1 the product code change is good to go. Just some test re-org is needed.

jujoramos replied:
I'll rewrite the tests and push the changes again, thanks for the suggestions!.

jinmeiliao commented on geode-assembly/src/test/java/org/apache/geode/tools/pulse/AbstractPulseConnectivityTest.java:
```diff
@@ -15,26 +15,31 @@
 
-@Category(IntegrationTest.class)
-public class PulseVerificationTest {
+public abstract class AbstractPulseConnectivityTest {
```
We would like to get rid of the usage of abstract test classes in favor of rules and parameterized tests. I see further improvements on these tests. But if you don't mind, I can pull your changes in and make modifications on top of yours.

jujoramos replied:
Thanks for the tips, I couldn't find anything related to this intention within 
[Writing Tests](https://cwiki.apache.org/confluence/display/GEODE/Writing+tests) but I'm okay anyway.
I'll give it one more try in my next commit later today, change things as you please :-).

jinmeiliao commented on geode-pulse/src/test/java/org/apache/geode/tools/pulse/internal/PulseAppListenerTest.java:
```diff
@@ -0,0 +1,91 @@
+
+  @After
+  public void tearDown() {
+    if (repository != null) {
```
this looks familiar, can you use EmbeddedPulseRule for this?

jujoramos replied:
Yes, I thought about using the existing rule but it implies adding a dependency from `geode-web` to `geode-assembly` as the rule is defined there, or duplicating the rule source code, that's why I didn't do it. Which approach would be best?.

jinmeiliao commented:
I will pull this in the moment I got a green pipeline

jujoramos commented:
@jinmeiliao: thanks for reviewing this :-).
One more quick question and I'll leave you alone: do I need to change the status of [GEODE-3292](https://issues.apache.org/jira/browse/GEODE-3292) to `RESOLVED`?. Or that's done when the pull request is merged into `develop`?. Couldn't find a clear answer from [Code Contributions](https://cwiki.apache.org/confluence/display/GEODE/Code+contributions) and [Developer Workflow](https://cwiki.apache.org/confluence/display/GEODE/Developer+Workflow).
Cheers.
